File: services/config/providers/envOverride.ts
/**
 * EnvOverrideProvider (G1-3) — env-based 配置覆盖, priority 高于 settings.json.
 *
 * 新 env: MOSSEN_CONFIG_OVERRIDES (推荐, Mossen 命名规范)
 * 旧 env: MOSSEN_INTERNAL_FC_OVERRIDES (deprecated, GrowthBook 时代名字, 仍兼容读取)
 *
 * 当两个 env 同时存在: 新 env 优先, 警告 stderr 一次.
 *
 * env 内容 = JSON object, key 是 mossen.<domain>.<feature> 或 tengu_xxx (alias 解析).
 *   例: MOSSEN_CONFIG_OVERRIDES='{"mossen.analytics.eventBatchConfig": {"maxQueueSize": 100}}'
 */

import {
  PROVIDER_PRIORITY,
  type ConfigValueSource,
  type MossenConfigProvider,
  type ProviderResult,
} from '../types.js'
import { resolveAliasedKey } from '../aliasMap.js'

const NEW_ENV_NAME = 'MOSSEN_CONFIG_OVERRIDES'
const DEPRECATED_ENV_NAME = 'MOSSEN_INTERNAL_FC_OVERRIDES'

let parsedCache: { source: 'new' | 'deprecated'; data: Record<string, unknown> } | null = null
let parseAttempted = false
let deprecationWarned = false

function parseEnvOverrides(): Record<string, unknown> | null {
  if (parseAttempted) return parsedCache?.data ?? null
  parseAttempted = true

  const newRaw = process.env[NEW_ENV_NAME]
  const deprecatedRaw = process.env[DEPRECATED_ENV_NAME]

  if (newRaw && deprecatedRaw && !deprecationWarned) {
    process.stderr.write(
      `[mossen] Warning: both ${NEW_ENV_NAME} and ${DEPRECATED_ENV_NAME} set; using ${NEW_ENV_NAME} (${DEPRECATED_ENV_NAME} is deprecated).\n`,
    )
    deprecationWarned = true
  } else if (!newRaw && deprecatedRaw && !deprecationWarned) {
    process.stderr.write(
      `[mossen] Warning: ${DEPRECATED_ENV_NAME} is deprecated; rename to ${NEW_ENV_NAME}.\n`,
    )
    deprecationWarned = true
  }

  const raw = newRaw ?? deprecatedRaw
  if (!raw) return null

  try {
    const parsed = JSON.parse(raw) as unknown
    if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
      const rawData = parsed as Record<string, unknown>
      // Alias-resolve env keys at parse time so callers can use either tengu_*
      // (legacy) or mossen.* (new) names; facade lookup is always alias-resolved.
      const data: Record<string, unknown> = {}
      for (const [k, v] of Object.entries(rawData)) {
        data[resolveAliasedKey(k)] = v
      }
      parsedCache = { source: newRaw ? 'new' : 'deprecated', data }
      return data
    }
    process.stderr.write(
      `[mossen] Warning: ${NEW_ENV_NAME}/${DEPRECATED_ENV_NAME} must be a JSON object, got ${typeof parsed}; ignoring.\n`,
    )
    return null
  } catch (e) {
    process.stderr.write(
      `[mossen] Warning: failed to parse ${NEW_ENV_NAME}/${DEPRECATED_ENV_NAME}: ${(e as Error).message}; ignoring.\n`,
    )
    return null
  }
}

/** 仅供测试: 重置内部 cache */
export function _resetEnvOverrideCacheForTesting(): void {
  parsedCache = null
  parseAttempted = false
  deprecationWarned = false
}

export class EnvOverrideProvider implements MossenConfigProvider {
  readonly name: ConfigValueSource = 'env'

  readonly priority = PROVIDER_PRIORITY.env

  readonly enabled = true

  get<T>(key: string): ProviderResult<T> {
    const data = parseEnvOverrides()
    if (data && Object.prototype.hasOwnProperty.call(data, key)) {
      return { value: data[key] as T, source: 'env' }
    }
    return undefined
  }
}

File: crates/mossen-agent/src/services/config/providers/env_override.rs
//! EnvOverrideProvider — env-based configuration override (priority 1).
//!
//! Reads MOSSEN_CONFIG_OVERRIDES (preferred) or MOSSEN_INTERNAL_FC_OVERRIDES (deprecated).
//! Content is a JSON object with mossen.* or tengu_* keys.

use once_cell::sync::Lazy;
use parking_lot::Mutex;
use serde_json::Value;
use std::collections::HashMap;
use std::env;

use crate::services::config::alias_map::resolve_aliased_key;
use crate::services::config::types::{ConfigValueSource, MossenConfigProvider, ProviderResult};

const NEW_ENV_NAME: &str = "MOSSEN_CONFIG_OVERRIDES";
const DEPRECATED_ENV_NAME: &str = "MOSSEN_INTERNAL_FC_OVERRIDES";

struct EnvOverrideCache {
    data: Option<HashMap<String, Value>>,
    parse_attempted: bool,
    deprecation_warned: bool,
}

static CACHE: Lazy<Mutex<EnvOverrideCache>> = Lazy::new(|| {
    Mutex::new(EnvOverrideCache {
        data: None,
        parse_attempted: false,
        deprecation_warned: false,
    })
});

fn parse_env_overrides() -> Option<HashMap<String, Value>> {
    let mut cache = CACHE.lock();
    if cache.parse_attempted {
        return cache.data.clone();
    }
    cache.parse_attempted = true;

    let new_raw = env::var(NEW_ENV_NAME).ok();
    let deprecated_raw = env::var(DEPRECATED_ENV_NAME).ok();

    if new_raw.is_some() && deprecated_raw.is_some() && !cache.deprecation_warned {
        eprintln!(
            "[mossen] Warning: both {} and {} set; using {} ({} is deprecated).",
            NEW_ENV_NAME, DEPRECATED_ENV_NAME, NEW_ENV_NAME, DEPRECATED_ENV_NAME
        );
        cache.deprecation_warned = true;
    } else if new_raw.is_none() && deprecated_raw.is_some() && !cache.deprecation_warned {
        eprintln!(
            "[mossen] Warning: {} is deprecated; rename to {}.",
            DEPRECATED_ENV_NAME, NEW_ENV_NAME
        );
        cache.deprecation_warned = true;
    }

    let raw = new_raw.or(deprecated_raw);
    let raw = match raw {
        Some(r) => r,
        None => return None,
    };

    match serde_json::from_str::<Value>(&raw) {
        Ok(Value::Object(map)) => {
            let mut data = HashMap::new();
            for (k, v) in map {
                let resolved = resolve_aliased_key(&k).to_string();
                data.insert(resolved, v);
            }
            cache.data = Some(data.clone());
            Some(data)
        }
        Ok(_) => {
            eprintln!(
                "[mossen] Warning: {}/{} must be a JSON object; ignoring.",
                NEW_ENV_NAME, DEPRECATED_ENV_NAME
            );
            None
        }
        Err(e) => {
            eprintln!(
                "[mossen] Warning: failed to parse {}/{}: {}; ignoring.",
                NEW_ENV_NAME, DEPRECATED_ENV_NAME, e
            );
            None
        }
    }
}

/// Reset internal cache (testing only).
pub fn reset_env_override_cache_for_testing() {
    let mut cache = CACHE.lock();
    cache.data = None;
    cache.parse_attempted = false;
    cache.deprecation_warned = false;
}

/// Environment variable override provider.
pub struct EnvOverrideProvider;

impl MossenConfigProvider for EnvOverrideProvider {
    fn name(&self) -> ConfigValueSource {
        ConfigValueSource::Env
    }

    fn priority(&self) -> u8 {
        ConfigValueSource::Env.priority()
    }

    fn enabled(&self) -> bool {
        true
    }

    fn get(&self, key: &str) -> Option<ProviderResult> {
        let data = parse_env_overrides()?;
        data.get(key).map(|value| ProviderResult {
            value: value.clone(),
            source: ConfigValueSource::Env,
            resolved_key: None,
        })
    }
}
